[AIRFLOW-5448] Handle istio-proxy for Kubernetes Pods

sjmiller609 opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW-5448) issues and references them in the PR title.

### Description

- [x] Here are some details about my PR:

Istio service mesh is not compatible by default with Kubernetes Jobs. The normal behavior is that a Job will be started, get an istio-proxy sidecar attached to it via the istio mutating webhook, run until completion, then the 'main' container in the pod stops, but istio-proxy hangs around indefinitely. This applies to the Kubernetes Executor.

Very recently, Istio implemented an endpoint that can be called to cleanly exit the proxy, specifically designed for this use case.

explanation: https://github.com/istio/istio/issues/15041
istio PR implementing it: https://github.com/istio/istio/pull/15406
Astronomer will make a contribution to handle cleanly exit the istio-proxy by default. This will help Astronomer and other Airflow users making use of the Kubernetes Executor in combination with Istio.

Here a PR I am closing in favor of contributing to upstream: https://github.com/astronomer/airflow/pull/47

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

 - [x] All the public functions and the classes in the PR contain docstrings that explain what it does

BasPH committed:
[AIRFLOW-4835] Refactor operator render_template (#5461)

- Refactors `BaseOperator.render_template()` and removes `render_template_from_field()`. The functionality could be greatly simplified into a single `render_template()` function.
- Improves performance by removing two `hasattr` calls and avoiding recreating Jinja environments.
- Removes the argument `attr` to `render_template()` which wasn't used.
- Squashes multiple similar tests into two parameterized tests.
- Adheres to 110 line length.
- Adds support for templating sets.
- Adds Pydoc.
- Adds typing.

(cherry picked from commit 47dd4c99a7324cb52121a25001599d4e7597959b)

danfrankj committed:
[AIRFLOW-5210] Make finding template files more efficient (#5815)

For large DAGs, iterating over template fields to find template files can be time intensive.
Save this time for tasks that do not specify a template file extension.

(cherry picked from commit eeac82318a6440b2d65f9a35b3437b91813945f4)

mik-laj commented on airflow/contrib/kubernetes/istio.py:
```diff
@@ -0,0 +1,148 @@
+    def _shutdown_istio_proxy(self, pod):
+        """Shutdown the istio-proxy on the provided pod
+
+        Args:
```
We use `:param:` style syntax of docstring:
https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html

sjmiller609 committed:
[AIRFLOW-5448] Handle istio-proxy for Kubernetes Pods

Istio service mesh is not compatible by default with Kubernetes Jobs.
The normal behavior is that a Job will be started, get an istio-proxy
sidecar attached to it via the istio mutating webhook, run until
completion, then the 'main' container in the pod stops, but istio-proxy
hangs around indefinitely. This change handles cleanly exiting the
Istio sidecar 'istio-proxy' when a Kubernetes Executor task completes.

sjmiller609 commented on setup.py:
```diff
@@ -226,6 +226,7 @@ def write_version(filename=os.path.join(*["airflow", "git_version"])):
             'thrift_sasl>=0.2.0',
             'snakebite[kerberos]>=2.7.8']
 kubernetes = ['kubernetes>=3.0.0',
+              'packaging>=19.1',
```
build is failing because this guy is missing on import.

sjmiller609 replied:
ok it's because this is not in python2

sjmiller609 replied:
i'll add the support in. one moment...

andrewldesousa committed:
[AIRFLOW-5449] Users can now set existing variable values to None (#6084)

mik-laj committed:
[AIRFLOW-5480] Fix flaky impersonation (#6098)


(cherry picked from commit 4459592e523bbeeda0747dd9f888c3e215a241d1)

potiuk committed:
[AIRFLOW-5256] Related pylint changes for common licences in python files (#5786)

  (cherry picked from commit 47801057989046dfcf7b424ce54afee103803815)

dimberman committed:
[AIRFLOW-5447] Scheduler stalls because second watcher thread in default args

potiuk committed:
[AIRFLOW-5369] Adds interactivity to pre-commits (#5976)

This commit adds full interactivity to pre-commits. Whenever you run pre-commit
and it detects that the image should be rebuild, an interactive question will
pop up instead of failing the build and asking to rebuild with REBUILD=yes

This is much nicer from the user perspective. You can choose whether to:
1) Rebuild the image (which will take some time)
2) Not rebuild the image (this will use the old image with hope it's OK)
3) Quit.

Answer to that question is carried across all images needed to rebuild.
There is the special "build" pre-commit hook that takes care about that.

Note that this interactive question cannot be asked if you run only
single pre-commit hook with Dockerfile because it can run multiple processes
and you can start building in parallel. This is not desired so instead we fail
such builds.

(cherry picked from commit 857788e305bbefe4566a1988e2072a21c7aab319)

kaxil committed:
[AIRFLOW-XXX] Add example of running pre-commit hooks on single file (#6143)

Or a list of files

yimingl17 committed:
[AIRFLOW-5519] Fix sql_to_gcs operator missing multi-level default args by adding apply_defaults decorator  (#6146)

(cherry picked from commit d313d8d24b1969be9154b555dd91466a2489e1c7)

Commit:
[AIRFLOW-5521] Fix link to GCP documentation (#6150)

(cherry picked from commit 7b6045a479edca06732fd766ef8ceabf41c7e82a)

potiuk committed:
[AIRFLOW-XXX] Update documentation about variables forcing answer (#6158)


(cherry picked from commit 04da038100786da179f977df4eb2bab5352949d9)

kaxil committed:
[AIRFLOW-XXX] Fix incorrect backticks in BREEZE.rst (#6161)


(cherry picked from commit e68d92f0899efa6b99bd1f89325e8a10fbbc8c63)

kaxil committed:
[AIRFLOW-XXX] Fix Prerequisites link in BREEZE.rst (#6160)


(cherry picked from commit 2d668ca9f280946cb1800af4ec99ad6727c8e363)

Commit:
[AIRFLOW-5526] Update docs configuration due to migration of GCP docs (#6154)

* [AIRFLOW-5526] Update docs configuration due to migration of GCP docs

mik-laj committed:
[AIRFLOW-XXX] Add a third way to configure authorization (#6134)


(cherry picked from commit 38ebc425fc8385e1493d1bdf7151affa16c844e2)

kaxil committed:
[AIRFLOW-XXX] Fix backtick issues in .rst files & Add Precommit hook (#6162)

(cherry picked from commit cfd8d605d9e29d06ff7066bded6c5d2c479d202e)

kaxil committed:
[AIRFLOW-5531] Replace deprecated log.warn() with log.warning() (#6165)

- Add pre-commit hook to detect this

(cherry picked from commit ccf44086c55b27325bfdd9e76ad1698f70d35cde)

potiuk committed:
[AIRFLOW-5533] Fixed failing CRON build

This change optimises further image building and removes unnecessary
verbosity in building the images for CI builds.

After this change is merged, only the necessary images are built for
each type of check:
* Tests -> only CI
* Static checks (with/without pylint) -> Only CI_SLIM
* Docs -> only CI_SLIM
* Licence checks -> Only CHECKLICENCE

Previously the right images only were built in ci_before_install.sh
but then in case of static checks, the pre-commit build image step
also rebuilt CHECKLICENCE and CI images - which was not necessary
and very long in case of CRON job - this caused the CRON job to
fail at 10m of inactivity.

potiuk committed:
[AIRFLOW-5535] Fix name of VERBOSE parameter

potiuk committed:
[AIRFLOW-5536] Better handling of temporary output files

potiuk committed:
[AIRFLOW-5537] Yamllint is not needed as dependency on host

It used to be needed for pre-commits but is not needed any more
as it is automatically installed as dependency in the virtualenv
created by pre-commit
